boot/freeldr/bootsect/ext.S

```diff
@@ -0,0 +1,717 @@
+
+    cmp si, dx                                              // End of buffer reached?
+    jae .Done                                               // Abort the search if yes
+
+    test eax, eax                                           // Check if Inode is zero
+    jz .Skip                                                // Skip this entry if yes
+
+    push si                                                 // Save SI
+    add si, EXT_DIRECTORY_ENTRY_NAME_OFFSET                 // Move to filename in entry
+    mov di, offset ExtLdrFileName                           // Load target filename address
+    mov cx, offset ExtLdrFileNameEnd - ExtLdrFileName       // Length of filename to compare
+    repe cmpsb                                              // Compare filenames
+    pop si                                                  // Restore SI
+    jz LoadExtLdr                                           // Found matching file
+
+.Skip:
+    // Move to next directory entry and continue looping
+    add si, word ptr ds:[si + EXT_DIRECTORY_ENTRY_SIZE_OFFSET]
+    jmp .FLoop
+.Done:
+    ret
+
+LoadExtLdr:
+    // Swap ES and DS
+    call SwapESWithDS
+
+    push si                                                 // Save SI
+    mov si, offset msgLoadingExtLdr                         // Point SI to a loading message
+    call PutChars                                           // Show the message
+    pop si                                                  // Restore SI
+
+    mov eax, dword ptr es:[si]                              // Load directory Inode
+    call ExtReadInode                                       // Read the inode
+    mov si, word ptr BP_REL(ExtReadInodeIndexOffset)        // Set the correct offset
+
+    // Get Inode type
+    mov ax, word ptr es:[si]
+    and ax, EXT_INODE_TYPE_MASK
+
+    cmp ax, EXT_INODE_TYPE_REGULAR                          // Check if regular file
+    jnz ExtLdrPrintRegFileError                             // If not, handle error
+
+    call BasicReadFile                                      // Load the file using basic function
+    call ExtSetOldFileSegment                               // Set old file segment
+    call SwapESWithDS                                       // Swap ES with DS before copy
+
+    // Copy the loaded file to 1KB ahead of this bootsector
+    xor si, si
+    mov di, offset ExtLdrEntryPoint
+    mov cx, EXTLDR_BOOTSECTOR_SIZE
+    rep movsb
+
+    ljmp16 0, ExtLdrEntryPoint
+
+ExtLdrPrintFileNotFound:
+    // Make DS correct
+    call SwapESWithDS
+
+    mov si, offset msgExtLdr
+    jmp DisplayItAndReboot
+ExtLdrPrintRegFileError:
+    mov si, offset msgExtLdrNotRegularFile      // ExtLdr not found message
+DisplayItAndReboot:
+    call PutChars                               // Display it
+    jmp Reboot
+
+msgExtLdr:
+    .ascii "extldr.sys not found", CR, LF, NUL
+msgExtLdrNotRegularFile:
+    .ascii "extldr.sys is not a regular file", CR, LF, NUL
+ExtLdrFileName:
```

Comment:
I would suggest you move this variable either before or after the error messages.